tensorflow_similarity/losses/softnn_loss.py

```diff
@@ -0,0 +1,74 @@
+from typing import Any, Dict, Union, Optional
+
+import tensorflow as tf
+from tensorflow.keras.losses import Loss
+from tensorflow_similarity.distances import Distance, distance_canonicalizer
+
+from tensorflow_similarity.types import FloatTensor, IntTensor
+
+
+@tf.keras.utils.register_keras_serializable(package="Similarity")
+class SoftNearestNeighborLoss(Loss):
```

Comment:
Let's make this a subclass of MetricLoss (see multisim_loss.py). We're going to work on a small refactor of the MetricLoss base class to better support multi GPU training and it would good to have all of these follow the same format.